lib/pbench/server/db/models/tracker.py

```diff
@@ -0,0 +1,638 @@
+            )
+
+        # TODO: this would be a good place to generate an audit log
+
+        self.state = new_state
+        self.transition = datetime.datetime.now()
+        self.update()
+
+    def add(self):
+        """
+        add Add the Dataset object to the database
+        """
+        try:
+            Database.db_session.add(self)
+            Database.db_session.commit()
+        except Exception as e:
+            self.logger.error("Can't add {} to DB", str(self))
+            Database.db_session.rollback()
+            raise DatasetSqlError(f"Error adding {self}: {e}")
+
+    def update(self):
+        """
+        update Update the database row with the modified version of the
+        Dataset object.
+        """
+        try:
+            Database.db_session.commit()
+        except Exception as e:
+            self.logger.error("Can't update {} in DB", str(self))
+            Database.db_session.rollback()
+            raise DatasetSqlError(f"Error updating {self}: {e}")
+
+
+class Metadata(Database.Base):
+    """ Retain secondary information about datasets
+
+    Columns:
+        id          Generated unique ID of table row
+        dataset_ref Dataset row ID (foreign key)
+        key         Metadata key string
+        value       Metadata value string
+    """
+
+    __tablename__ = "metadata"
+
+    # Currently defined metadata keys
+    REINDEX = "REINDEX"
+    ARCHIVED = "ARCHIVED"
+    TARBALL_PATH = "TARBALL_PATH"
+
+    METADATA_KEYS = [REINDEX, ARCHIVED, TARBALL_PATH]
+
+    id = Column(Integer, primary_key=True, autoincrement=True)
+    key = Column(String(255), unique=False, nullable=False, index=True)
+    value = Column(String(2048), unique=False, nullable=True)
+    dataset_ref = Column(
+        Integer, ForeignKey("dataset.id", ondelete="CASCADE"), nullable=False
+    )
+
+    def __init__(self, **kwargs):
+        """
+        Construct a new Metadata object for a dataset
+
+        Parameters:
+
+            key     Metadata key (from METADATA_KEYS)
+            value   String value for key
+        """
+        try:
+            key = kwargs.get("key")
+        except KeyError:
+            raise MetadataBadKey("Metadata key is required")
+        else:
+            if key not in Metadata.METADATA_KEYS:
+                raise MetadataBadKey(f"Metadata key {key} is not supported")
+        try:
+            value = kwargs.get("value")
+        except KeyError:
+            raise MetadataBadKey(f"Metadata key {key} value is required")
+        super().__init__(**kwargs)
+        self.key = key
+        self.value = value
+
+    @staticmethod
+    def create(**kwargs):
+        try:
+            meta = Metadata(**kwargs)
+            meta.add(kwargs.get("dataset"))
+        except KeyError:
+            raise MetadataError("Metadata create missing argument 'dataset'")
+        except Exception:
+            return None
```

Comment:
Dave, I'm too distracted/stupid right now to figure this out...what is expected to raise the `KeyError`?   Is it the `kwargs.get()` call?  (If so, there's another bug, here....)

Reply:
Same as your previous catch; this used to be `kwargs["dataset"]` and I left the `except` when I changed it.

What bothers me most, actually, is that I thought I had these covered in the unit tests; but since I didn't get any failures, apparently I missed that.

Reply:
I had the same thought...glad you reached it first.  ;-)